### ISSUE TITLE
# Milestone 17 — Extract Keras Models into a Versioned Python Library (`trading-models`)

### ISSUE THREAD

**Issue by khmusayev** [khmusayev/signal-inference#28]
> **Superseded**: This milestone has been split into M17a, M17b, M17c (see `m17a-steps.md`, `m17b-steps.md`, `m17c-steps.md`). This file is kept as the original design reference.

## Context

The ml-inference app currently embeds the LSTM model file, all preprocessing (normalization, time features), and inference logic directly in `app/signals/binance_btcusdt_1m_lstm.py`. This:
- Pollutes the main app with strategy-specific details
- Exposes the strategy in ml-inference's GitHub repo
- Makes it hard to version and update models independently

**Goal**: Move all model-specific code into a separate library (`trading-models`) that lives in the private `trader` repo (`C:\Users\yulii\development\trader`). The ml-inference app provides candles in, gets predictions out.

**Initial models** (confirmed solid):
- `lstm_indicators_long_model` — BTC/USDT 1min, 2-class (no_trade/long), 15 features, window=240
- `lstm_long_model_crossentropy` — BTC/USDT 1h, 2-class (no_trade/long), 9 features, window=480

The current `timestamp_3class` model in ml-inference can serve as a simpler reference for the extraction pattern.

---

### Location

```
trader/                          # existing private repo
  src/                           # training/backtesting code (unchanged)
  libs/
    trading-models/              # NEW — the publishable library
      pyproject.toml
      src/
        trading_models/
          __init__.py            # public API
          _registry.py           # name -> class mapping
          base.py                # ModelPredictor ABC
          candle.py              # Candle dataclass (library's own)
          prediction.py          # Prediction + Direction
          models/
            __init__.py          # imports all models to trigger registration
            indicators_long.py   # lstm_indicators_long_model (BTCUSDT 1min)
            long_crossentropy.py # lstm_long_model_crossentropy (BTCUSDT 1h)
          weights/
            indicators_long_btcusdt_1m.keras
            indicators_long_btcusdt_1m_scaler.joblib
            long_crossentropy_btcusdt_1h.keras
      tests/
        conftest.py
        test_indicators_long.py
        test_long_crossentropy.py
        test_registry.py
```

CI config lives at repo-level (`trader/.github/workflows/trading-models-ci.yml`) scoped via path filter.

### Public API

```python
from trading_models import get_predictor, list_predictors, ModelPredictor, Candle, Prediction, Direction

# At app startup (loads .keras once):
predictor = get_predictor("indicators_long")

# On each candle close (cheap — numpy + model.predict):
prediction = predictor.predict(candles)  # list[Candle] -> Prediction
prediction.direction      # Direction.LONG / NO_TRADE
prediction.probabilities  # raw softmax output, e.g. (0.8, 0.2)
```

### Key Types

**`candle.py`** — Library's own Candle (same fields as ml-inference's, avoids coupling):
```python
@dataclass(frozen=True)
class Candle:
    open_time: datetime
    open: float
    high: float
    low: float
    close: float
    volume: float
```

**`prediction.py`** — Output type:
```python
class Direction(IntEnum):
    SHORT = -1
    NO_TRADE = 0
    LONG = 1

@dataclass(frozen=True)
class Prediction:
    direction: Direction
    probabilities: tuple[float, ...]
```

**`base.py`** — ABC that all model variants implement:
```python
class ModelPredictor(ABC):
    @property
    def name(self) -> str: ...
    @property
    def window_size(self) -> int: ...
    @property
    def num_classes(self) -> int: ...
    def predict(self, candles: list[Candle]) -> Prediction: ...
```

**`_registry.py`** — Factory + decorator registration:
```python
@register("indicators_long")
class IndicatorsLongPredictor(ModelPredictor): ...

# Usage:
predictor = get_predictor("indicators_long")
```

#### `indicators_long.py` — lstm_indicators_long_model (BTCUSDT 1min)

Source: `trader/src/ml_gpu/lstm_indicators_long_model.py`

**Properties**: `name="indicators_long"`, `window_size=240`, `num_classes=2`

**Features (15 total per candle)**:
- OHLCV (5): open, high, low, close, volume
- Time features (4): sin_time, cos_time, sin_day, cos_day
- Indicator features (6), pre-scaled via MaxAbsScaler:
  1. `vol_exp_norm` = (ATR_fast(5) - ATR_slow(20)) / ATR_slow(20)
  2. `trend_pos_norm` = (close / EMA(21)) - 1
  3. `trend_slope_norm` = (EMA(21) - EMA(21)[lag=5]) / ATR_fast(5)
  4. `rel_vol_norm` = log(volume / SMA(20, volume))
  5. `previous_session_max` = (close - prev_day_high) / close
  6. `previous_session_min` = (close - prev_day_low) / close

**Normalization**: OHLC z-score together, volume z-score separately, time features unchanged, indicator features pre-scaled by `MaxAbsScaler` (fitted on training data).

**Important**: The fitted `MaxAbsScaler` must be saved during training (`joblib.dump(scaler, "scaler.joblib")`) and loaded at predictor init. This is a one-time training script change.

**Previous-session features challenge**: Features 5-6 require the previous calendar day's high/low. At inference time, the predictor's `predict()` must compute these from the candle window. This requires the window to contain candles spanning at least into the previous day. With `window_size=240` (4 hours of 1-min candles), this is **not guaranteed**. Options:
- (a) Increase `window_size` to 1680 (240 + 1440) to always cover the previous day — wasteful.
- (b) Accept `previous_day_high` and `previous_day_low` as optional extra inputs to `predict()`.
- (c) Require the caller to pass enough candles for indicator warm-up (the cleaner option — document the minimum as `window_size + indicator_warmup`).

**Decision**: Option (b) — add an optional `context` dict parameter to `predict()` for extra data that can't be derived from the candle window alone. This keeps the base API clean while allowing model-specific extensions.

**Output**: 2-class softmax — `[no_trade_prob, long_prob]`. `argmax == 1` → Direction.LONG, else Direction.NO_TRADE.

#### `long_crossentropy.py` — lstm_long_model_crossentropy (BTCUSDT 1h)

Source: `trader/src/ml_gpu/btcusdt_1h/lstm_long_model_crossentropy.py`

**Properties**: `name="long_crossentropy"`, `window_size=480`, `num_classes=2`

**Features (9 total per candle)**:
- OHLCV (5): open, high, low, close, volume
- Time features (4): sin_time, cos_time, sin_day, cos_day

**Normalization**: Identical to the current timestamp model — OHLC z-score together, volume z-score separately, time features unchanged. No indicators, no scaler needed.

**Output**: 2-class softmax — `[no_trade_prob, long_prob]`. Same mapping as above.

**Note**: This model uses the same `get_time_features()` and `normalize_ohlcv_time()` functions as the current production model. These can be shared as private utilities within the library (e.g., `trading_models/_preprocessing.py`).

### What stays in ml-inference

- **SL/TP calculation** — `calculate_stop_loss()` and `calculate_take_profit()` move to `app/signals/risk.py`. These are trade-execution policy (2:1 R:R, min-low/max-high), not model-specific.
- **Sliding window management** — `deque(maxlen=window_size)`, candle gap validation, seeding from broker REST API
- **Signal construction** — combining prediction + SL/TP into a `Signal` dataclass
- **All candle collector generators** for other brokers (unchanged)

### `pyproject.toml`

```toml
[project]
name = "trading-models"
version = "0.1.0"
requires-python = ">=3.12,<3.13"
dependencies = ["numpy", "tensorflow>=2.16.1,<2.19"]

[project.optional-dependencies]
indicators = ["TA-Lib"]
dev = ["pytest"]

[tool.setuptools.packages.find]
where = ["src"]

[tool.setuptools.package-data]
trading_models = ["weights/*.keras", "weights/*.joblib"]

[build-system]
requires = ["setuptools>=68.0", "wheel"]
build-backend = "setuptools.build_meta"
```

---

### New file: `app/signals/model_signal_generator.py`

A generic signal generator that wraps any `ModelPredictor`. Seeding is broker-agnostic — the generator receives a `seed_fn` callable that returns historical candles:

```python
class ModelSignalGenerator(SignalGenerator):
    def __init__(self, predictor, seed_fn, broker, symbol, timeframe, expected_gap):
        self._predictor = predictor          # already loaded, predict() is cheap
        self._seed_fn = seed_fn              # callable() -> list[Candle]
        self._window_size = predictor.window_size
        self._candles: deque[Candle] = deque(maxlen=self._window_size)
        self._broker = broker
        self._symbol = symbol
        self._timeframe = timeframe
        self._expected_gap = expected_gap

    def seed(self):
        candles = self._seed_fn()
        self._candles = deque(candles, maxlen=self._window_size)

    async def on_candle(self, candle):
        # sliding window + gap check (same as current)
        self._candles.append(candle)

        # convert to library's Candle type
        from trading_models import Candle as TMCandle
        tm_candles = [
            TMCandle(c.open_time, c.open, c.high, c.low, c.close, c.volume)
            for c in self._candles
        ]
        prediction = self._predictor.predict(tm_candles)
        # build Signal with SL/TP from risk module
```

**Candle conversion note**: ml-inference's `Candle` and the library's `Candle` have identical fields. The conversion is a trivial positional copy. Alternatively, ml-inference could adopt `trading_models.Candle` as its canonical type to avoid this, but that couples core types to the library. The positional copy is ~0.01ms for 480 candles — negligible.

### New file: `app/signals/risk.py`

Move `calculate_stop_loss()` and `calculate_take_profit()` here (from `binance_btcusdt_1m_lstm.py` lines 164-185).

### Modified: `app/session/manager.py`

- Remove `MODEL_PATH` constant
- Remove `BinanceBtcUsdt1mLstmSignalGenerator` import
- Add `from trading_models import get_predictor`
- Add model mapping table:

```python
MODEL_MAPPING: dict[tuple[str, str, str], str] = {
    ("binance", "BTCUSDT", "1m"): "indicators_long",
    ("binance", "BTCUSDT", "1h"): "long_crossentropy",
}
```

- In `_create_generator()`: look up `MODEL_MAPPING.get((broker, symbol, timeframe))`. If found, create `ModelSignalGenerator` with the matching predictor. If not, fall through to candle collector generators as before.
- Predictors are instantiated once and cached (e.g., in a `_predictors` dict on `SessionManager`), so the `.keras` file is loaded only once even if the stream is stopped and restarted.

### Deleted: `app/signals/binance_btcusdt_1m_lstm.py`

Fully replaced by `model_signal_generator.py` + `risk.py` + library.

### Deleted: `model/lstm_timestamp_model_jun_oct.keras`

No longer needed — models are bundled inside the library package.

### Modified: `requirements.txt`

Add:
```
trading-models @ git+ssh://git@github.com/USERNAME/trader.git@trading-models-v0.1.0#subdirectory=libs/trading-models
```

### Modified: tests

- `tests/signals/test_helpers.py` -> update imports to `app.signals.risk`
- `tests/signals/test_binance_btcusdt_1m_lstm.py` -> rewrite as `test_model_signal_generator.py` testing the generic generator with a mock `ModelPredictor`

---

## Distribution

**Recommended: `pip install git+ssh://...` with git tags.** Simple, no registry needed.

- Version pinned via git tag (e.g., `trading-models-v0.1.0`)
- `#subdirectory=libs/trading-models` tells pip where the package is in the monorepo
- SSH key on the machine running ml-inference (already needed for git clone)

Local dev: `pip install -e C:\Users\yulii\development\trader\libs\trading-models`

### Docker build with private repo

The Dockerfile needs SSH access to clone from the private `trader` repo during `pip install`. Use Docker BuildKit SSH forwarding:

```dockerfile
# syntax=docker/dockerfile:1
FROM python:3.12-slim

# Install system deps (including libta-lib for indicator models)
RUN apt-get update && apt-get install -y --no-install-recommends \
    libta-lib-dev git openssh-client && rm -rf /var/lib/apt/lists/*

COPY requirements.txt .
RUN --mount=type=ssh pip install --no-cache-dir -r requirements.txt

COPY . .
CMD ["uvicorn", "app.main:app", "--host", "0.0.0.0", "--port", "8000"]
```

Build with: `DOCKER_BUILDKIT=1 docker build --ssh default .`

On CI/CD (GitHub Actions), use a deploy key:
1. Generate an SSH key pair
2. Add the public key as a deploy key on the `trader` repo (read-only)
3. Add the private key as a secret on the `ml-inference` repo
4. In the workflow, configure `ssh-agent` before `docker build`

---

## CI for the Library

Add to `trader/.github/workflows/trading-models-ci.yml`:

```yaml
name: trading-models CI
on:
  push:
    branches: [main]
    paths: ['libs/trading-models/**']
  pull_request:
    branches: [main]
    paths: ['libs/trading-models/**']
jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-python@v5
        with: { python-version: '3.12' }
      - name: Install TA-Lib system library
        run: |
          sudo apt-get update
          sudo apt-get install -y libta-lib-dev
      - name: Install package with dev and indicators deps
        working-directory: libs/trading-models
        run: pip install -e ".[dev,indicators]"
      - name: Run tests
        working-directory: libs/trading-models
        run: pytest tests/ -v
```

---

## Adding a New Model (End-to-End)

1. Train in `trader/src/ml_gpu/` as usual
2. Create `libs/trading-models/src/trading_models/models/new_model.py` with `@register("new_model")` decorator
3. Copy `.keras` file (and `scaler.joblib` if applicable) to `weights/`
4. Import in `models/__init__.py`
5. Write tests
6. Bump version in `pyproject.toml`, push, tag `trading-models-v0.2.0`
7. Update `ml-inference/requirements.txt` to new tag
8. In `ml-inference/app/session/manager.py`, add entry to `MODEL_MAPPING`

---

## Model Loading — Addressing the Core Concern

**Yes, the model loads only once.** The `ModelPredictor.__init__()` calls `keras.models.load_model()` (expensive, ~1-3s). The `predict()` method only does numpy preprocessing + `model.predict()` (cheap, ~10-50ms). The predictor instance is created once in `SessionManager` and cached in `_predictors`, reused for the lifetime of the process. Updating the library version requires restarting the app (same as today with the bundled `.keras` file).

Keras lazy import (`from keras.models import load_model` inside `__init__`) is preserved so that importing `trading_models` doesn't trigger TensorFlow initialization — only calling `get_predictor()` does.

---

## Implementation Order

1. **Create library scaffold** in `trader/libs/trading-models/` — pyproject.toml, package structure, base.py, registry, candle.py, prediction.py
2. **Extract model variants** — move preprocessing + inference code, copy `.keras` + `.joblib` files to weights/. For indicators model: update training script to also `joblib.dump(scaler)`.
3. **Write library tests** — test preprocessing functions, test predict with mocked keras model, test registry
4. **Add CI workflow** — `trader/.github/workflows/trading-models-ci.yml` (with TA-Lib system dep)
5. **Create ml-inference changes** — `model_signal_generator.py`, `risk.py`, update `manager.py` (with `MODEL_MAPPING`), update `requirements.txt`
6. **Update ml-inference tests** — adapt existing tests to new structure
7. **Update Dockerfile** — add SSH mount for private repo, add `libta-lib-dev`
8. **Tag and verify** — tag library version, install in ml-inference, run full test suite + Docker smoke test

## Verification

- **Library unit tests**: `cd trader/libs/trading-models && pytest tests/ -v`
- **ml-inference unit tests**: `cd ml-inference && pytest tests/ -v`
- **Integration**: `docker compose up --build` and `curl -X POST http://localhost:8000/session/binance/BTCUSDT/1m/start` — verify model loads and produces signals
- **Install test**: `pip install git+ssh://...#subdirectory=libs/trading-models` in a clean venv

## Key Challenges

1. **Indicator model's `MaxAbsScaler`**: The scaler was fitted during training. Must update the training script to `joblib.dump(scaler, "scaler.joblib")` and ship alongside the `.keras` file in `weights/`. One-time effort per model variant.
2. **Indicator model's previous-session features**: Require previous day's high/low which may not be derivable from the 240-candle window alone. Solved via optional `context` parameter on `predict()`.
3. **`.keras` files in git**: Moving from repo root to `libs/trading-models/src/trading_models/weights/`. Old root-level files should be `.gitignore`'d after migration.
4. **TensorFlow import time**: Lazy import pattern (importing keras inside `__init__`, not at module level) must be preserved.
5. **Docker SSH for private repo**: Requires BuildKit SSH mount + deploy key on CI. Documented above.
6. **TA-Lib system dependency**: Requires `libta-lib-dev` on Linux. Must be installed in both CI and Docker image.

**Comment by khmusayev** [comment]
https://github.com/khmusayev/signal-inference/pull/29
17a and 17b on trader repo
